### PR TITLE
Signed responses

### LINKED ISSUE #13: Signed responses
Implement signed server responses.
The server responds to the client with a signed response, which the client could store to attest to someone else that the response was indeed emitted from the trusted enclave.

This may be made optional so that clients that do not use it don't have to pay for it.

### PR DESCRIPTION
**Pull request by cchudant**
Closes #13 

Thanks to @CerineBnsd for the client part!

### CODE REVIEW

**Comment by JoFrost** [comment]
Thanks a lot! Will review asap

**Comment by JoFrost** [comment]
@CerineBnsd Can you use Black on the python code?

File: client/blindai/client.py [hunk 1]
```diff
@@ -300,3 +335,8 @@ def close_connection(self):
             self.channel = None
             self.stub = None
             self.policy = None
+
+    def export_proof(self):
+        """Dump the proof of execution to a json file"""
+        with open("execution_proof.json", "w") as proof_file:
```
Comment (JoFrost) [inline client/blindai/client.py:341]: Maybe the docstring can be updated a bit here, explaining the kind of data that will be exported.

**Review by JoFrost — COMMENTED** [review]
All good. Maybe the docstring part as I mentioned would need a few tweaks, but all the rest looks good.

**Comment by CerineBnsd** [comment]
> Thanks for formatting the code. Can anyone of you bump the client and version to 0.3.0, and update the gitbook documentation to reflect all those changes?

I'll update the docstrings, the versions, the client CHANGELOG, and interface reference in gitbook. I think it would be better if @Charles updates the parts that are related to the server as he may explain it better (especially the refactor as it isn't added yest as far as I know).Otherwise, I'll try to do it.

**Comment by JoFrost** [comment]
The refactor has already been documented! What is left to document is the new feature. You can do the client part and @cchudant the server part.

**Comment by CerineBnsd** [comment]
> The refactor has already been documented! What is left to document is the new feature. You can do the client part and @cchudant the server part.

It's done for the client-side.

**Comment by JoFrost** [comment]
> > The refactor has already been documented! What is left to document is the new feature. You can do the client part and @cchudant the server part.
> 
> It's done for the client-side.

Thanks! However, as not the whole documentation is ready and the pull request not done, can you push your changes to the doc as a pull request instead? This way we can merge the moment the feature is added for good to the main branch.

**Comment by CerineBnsd** [comment]
> Thanks! However, as not the whole documentation is ready and the pull request not done, can you push your changes to the doc as a pull request instead? This way we can merge the moment the feature is added for good to the main branch.

That's right, I didn't pay attention. I'll fix it.

**Comment by cchudant** [comment]
I've seen that the client now returns the output of the inference instead of a response object with an "output" field
Was this intended? @CerineBnsd 

@JoFrost We are currently pushing the signatures in an array, I was thinking about whether this is the best idea
There is currently no good way to get the signature of an individual signed call, maybe we want to return the proof in the response object returned by run_model and send_model. What do you think?

I will update the server side documentation

**Comment by CerineBnsd** [comment]
@cchudant For instance, the response object contains the signature + the output of the inference. As we're not returning the response containing the signature in upload_model, it's more consistent not to return the signature in run_model as well, and then what's left as information in the response object will be only the output array.

**Comment by JoFrost** [comment]
> I've seen that the client now returns the output of the inference instead of a response object with an "output" field Was this intended? @CerineBnsd
> 
> @JoFrost We are currently pushing the signatures in an array, I was thinking about whether this is the best idea There is currently no good way to get the signature of an individual signed call, maybe we want to return the proof in the response object returned by run_model and send_model. What do you think?
> 
> I will update the server side documentation

I definitely agree with your suggestion. @clauverjat What do you think?

**Comment by JoFrost** [comment]
@cchudant Let's apply your suggestion. Anyway we are only supposed to get the signature of the model we request, and not something else.

**Comment by JoFrost** [comment]
Alright, the protected branch rules works perfectly.

I will do a final review if you are done. We should expect to merge this on Friday if everything works well. Don't forget about the documentation for the server side as well

**Comment by JoFrost** [comment]
@cchudant The run tests CI failed again. I think the protobuf python files generation is not working anymore, according to the logs.

**Comment by cchudant** [comment]
how did that pass the CI, what
I'm so confused

File: server/CHANGELOG [hunk 2]
```diff
@@ -4,6 +4,11 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [0.3.0] ## 2022-03-18
+### Added
```
Comment (JoFrost) [inline server/CHANGELOG:8]: The addition/changes are missing.

**Review by JoFrost — CHANGES_REQUESTED** [review]
Add the changes in the changelog, and we can merge this.

File: server/CHANGELOG (hunk 2 above)
Comment (JoFrost) [reply server/CHANGELOG]: In order to win some times, I did the change myself. We can speak directly about what is coming next on monday.

**Review by JoFrost — APPROVED** [review]

**Pull request merged**